File: euler_095.py
from time import time

start = time()

sum_of_proper = [0] + [1] * 10 ** 6

for n in range(2, 5 * 10 ** 5 + 1):
    k = 2 * n
    while k < 10 ** 6 + 1:
        sum_of_proper[k] += n
        k += n

chains = {n:[n] for n in range(1, 10 ** 6 + 1)}
checked = [True, True] + [False] * (10 ** 6 - 1)

for n in range(2, 10 ** 6 + 1):

    if not checked[n]:
        chain = [n]

        m = sum_of_proper[n]

        if m != n and m <= 10 ** 6:
            chain.append(m)

        searching = True

        while searching:
            if m > 10 ** 6:
                searching = False
            else:
                m = sum_of_proper[m]

                if m in chain or m > 10 ** 6:
                    searching = False
                else:
                    chain.append(m)

        if m < 10 ** 6 + 1:
            k = chain.index(m)

            chain = chain[k:]

            for t in chain:
                chains[t] = chain
                checked[t] = True

chain_lengths = {n:len(chains[n]) for n in range(1, 10 ** 6 + 1)}

mc = max(chain_lengths.values())
smallest = min([n for n in range(1, 10 ** 6 + 1) if chain_lengths[n] == mc])

print("Smallest number and chain length: ", smallest, mc)

stop = time()
print("Time: " + str(stop - start))

File: euler_095.js
const start = Date.now();

// Proper Divisors

let sumOfProperDivisors = new Array(1000001);
sumOfProperDivisors.fill(0);
sumOfProperDivisors[1] = 1;
for (let div = 2; div < 1000001; div++) {
    let k = 2;
    while (k * div < 1000001) {
        sumOfProperDivisors[k * div] += div;
        k++;
    }
    sumOfProperDivisors[div] += 1;
}

// Tracking numbers we have already checked

let checked = new Array(10**6 + 2);
checked.fill(false);
checked[0] = checked[1] = checked[2] = true;

// Longest chain info

let smallest = 10**6 + 1;
let longest = 0;


// Building the chains

const chain = (num) => {
    let res = [num];
    let current = sumOfProperDivisors[res];

    while (current < 10**6 + 1 && !res.includes(current)) {
        res.push(current);
        current = sumOfProperDivisors[current];
    }

    for (const el of res) {
        if (el < 10**6 + 1) {
            checked[el] = true;
        }
    }

    if (res[res.length - 1] > 10**6) {
        return [];
    }

    let i = res.indexOf(current);
    res = res.slice(i);
    return res.slice(0,res.length);
}

for (let num = 3; num < 10**6 + 1; num++) {
    if (!checked[num]) {
        let c = chain(num).sort((a, b) => a - b);
        if (c.length > longest) {
            longest = c.length;
            smallest = c[0];
        }
    }
}


const end = Date.now();

console.log("Execute time: ", end - start, " The smallest member of the logest chain is: ", smallest);